src/helpers/loadOptimize.spec.ts

```diff
@@ -0,0 +1,68 @@
+import { receiveExperiments } from '../app/featureFlags/actions';
+import { Store } from '../app/types';
+import { assertDocument, assertWindow } from '../app/utils';
+import config from '../config';
+import createTestStore from '../test/createTestStore';
+import loadOptimize from './loadOptimize';
+
+describe('loadOptimize', () => {
+  let store: Store;
+  let window: Window;
+  let dispatch: jest.SpyInstance;
+  const document = assertDocument();
+
+  beforeEach(() => {
+    store = createTestStore();
+    window = assertWindow();
+    dispatch = jest.spyOn(store, 'dispatch');
+
+    const createEvent = document.createEvent.bind(document);
+    const originalCreateElement = document.createElement.bind(document);
+
+    document.createElement = jest.fn((type: string) => {
+      const element = originalCreateElement(type);
+      const event = createEvent('Event');
+      event.initEvent('load');
+      setImmediate(() => element.onload && element.onload(event));
+
+      return element;
+    });
+  });
+
+  afterEach(() => {
+    document.head.innerHTML = '';
+    window.dataLayer = [];
+  });
+
+  it('injects correct <script> into head if in production', async() => {
+    config.DEPLOYED_ENV = 'openstax.org';
+    window.dataLayer.push = jest.fn();
+
+    await loadOptimize(window, store);
+    const script = document.head.querySelector('script');
+
+    expect(script!.src).toEqual(
+      'https://www.googleoptimize.com/optimize.js?id=OPT-NFHSM4B'
+    );
+  });
+
+  it('injects correct <script> into head if in development', async() => {
+    config.DEPLOYED_ENV = 'foo';
+    window.dataLayer.push = jest.fn();
+
+    await loadOptimize(window, store);
+    const script = document.head.querySelector('script');
+
+    expect(script!.src).toEqual(
+      'https://www.googleoptimize.com/optimize.js?id=OPT-W65B3CP'
+    );
+  });
+
+  it('registers optimize callback and correctly dispatches action', async() => {
+    config.DEPLOYED_ENV = 'foo';
+    await loadOptimize(window, store);
+
+    ((window.dataLayer[0] as any)[2] as any).callback('1', 'OCCkMMCZSwW87szzpniCow');
+    expect(dispatch).toHaveBeenCalledWith(receiveExperiments(['OCCkMMCZSwW87szzpniCow', '1']));
```

Comment:
@TomWoodward is it ok to register sessions to dev here, or can you recommend a better way to test the optimize callback?

Reply:
Is it loading the actual script when you call `await loadOptimize`? Maybe we could just prevent this?

Reply:
to make it a unit test i think you'd mock the appendElement and asset the given script tag against a snapshot, you'd have to also mock the gtag and dataLayer calls, invoke the callback in the test with some known input, and then assert that dispatch is correctly called.

that is probably the right way to do it, although i'm a little sad about losing the confirmation that we're doing the calls correctly for the library because they are a little arcane, it is really better to not have the unit tests depending on loading external scripts

Reply:
updated